### PR TITLE
[FLINK-32326][state] Disable WAL in RocksDBWriteBatchWrapper by default.

### PR DESCRIPTION
**Pull request by StefanRRichter**
## What is the purpose of the change

Disables WAL by default in RocksDBWriteBatchWrapper for the case that now explicit WriteOption is passed in. This is the case in all restore operations and can impact the performance.

## Brief change log

- Creates default WriteOptions with WAL disabled in RocksDBWriteBatchWrapper.
- Unit tests in RocksDBWriteBatchWrapperTest

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change added tests and can be verified as follows:

Run RocksDBWriteBatchWrapperTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java [hunk 1]
```diff
@@ -125,10 +130,16 @@ public WriteOptions getOptions() {
...
-            flush();
+        try {
+            if (batch.count() != 0) {
+                flush();
+            }
+        } finally {
+            IOUtils.closeQuietly(batch);
+            if (ownsWriteOptions) {
+                IOUtils.closeQuietly(options);
+            }
```
[the remark this answers is not in the document]
Comment (StefanRRichter) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java:141]: Sorry if it's a stupid question, but what is "closer"? Like closable registry?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java [hunk 2]
```diff
@@ -79,7 +82,16 @@ public RocksDBWriteBatchWrapper(
...
+        if (options != null) {
+            this.options = options;
+            // ensure we don't close the object we don't own
+            this.ownsWriteOptions = false;
+        } else {
+            // Use default write options with disabled WAL
+            this.options = new WriteOptions().setDisableWAL(true);
+            // ensure we close the object we own
+            this.ownsWriteOptions = true;
+        }
```
[the remark this answers is not in the document]
Comment (StefanRRichter) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java:94]: I was starting to implement this in the way you suggested, but the backend WriteOptions are currently managed by `RocksDBResourceContainer` and calls to the getter of WriteOptions always create new object and the ownership is claimed by the container. If we don't want to create multiple objects that stick around until the backend is disposed, we would have to create the options before creating the backend and passing it in everywhere. Then the ownership would be pretty hard to determine along the way. I think this is actually the safest option for the change and the closest to the current code.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java]: ops! yes ;) I've used [guava](https://guava.dev/releases/19.0/api/docs/com/google/common/io/Closer.html) nomenclature

**Review by pnowojski — APPROVED** [review]
LGTM % one open nit, but feel free to ignore it if you like the current way better.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (StefanRRichter) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java]: done!

**Review by carp84 — APPROVED** [review]
+1. Maybe this worth a release note since `RocksDBWriteBatchWrapper` is a public class and might be used in some customized RocksDB state backends (just in case there's any downstream assumptions)

**Pull request merged**